proving_system/stark/src/cairo_vm/cairo_instruction.rs

```diff
@@ -0,0 +1,545 @@
+    type Error = InstructionDecodingError;
+
+    fn try_from(cell: &CairoMemoryCell) -> Result<Self, Self::Error> {
+        let flags = cell.value.limbs[3] >> FLAGS_OFFSET;
+        let op1_src = ((flags & OP1_SRC_MASK) >> OP1_SRC_OFF) as u8;
+
+        match op1_src {
+            0 => Ok(Op1Src::Op0),
+            1 => Ok(Op1Src::Imm),
+            2 => Ok(Op1Src::FP),
+            4 => Ok(Op1Src::AP),
+            _ => Err(InstructionDecodingError::InvalidOp1Src),
+        }
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum ResLogic {
+    Op1 = 0,
+    Add = 1,
+    Mul = 2,
+    // TODO: Check if this is correct
+    Unconstrained,
+}
+
+impl TryFrom<&CairoMemoryCell> for ResLogic {
+    type Error = InstructionDecodingError;
+
+    fn try_from(cell: &CairoMemoryCell) -> Result<Self, Self::Error> {
+        let flags = cell.value.limbs[3] >> FLAGS_OFFSET;
+        let res_logic = ((flags & RES_LOGIC_MASK) >> RES_LOGIC_OFF) as u8;
+
+        match res_logic {
+            0 => Ok(ResLogic::Op1),
+            1 => Ok(ResLogic::Add),
+            2 => Ok(ResLogic::Mul),
+            // TODO: Check this is correct
+            4 => Ok(ResLogic::Unconstrained),
+            _ => Err(InstructionDecodingError::InvalidResLogic),
+        }
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum PcUpdate {
+    Regular = 0,
+    Jump = 1,
+    JumpRel = 2,
+    Jnz = 4,
+}
+
+impl TryFrom<&CairoMemoryCell> for PcUpdate {
+    type Error = InstructionDecodingError;
+
+    fn try_from(cell: &CairoMemoryCell) -> Result<Self, Self::Error> {
+        let flags = cell.value.limbs[3] >> FLAGS_OFFSET;
+        let pc_update = ((flags & PC_UPDATE_MASK) >> PC_UPDATE_OFF) as u8;
+
+        match pc_update {
+            0 => Ok(PcUpdate::Regular),
+            1 => Ok(PcUpdate::Jump),
+            2 => Ok(PcUpdate::JumpRel),
+            4 => Ok(PcUpdate::Jnz),
+            _ => Err(InstructionDecodingError::InvalidPcUpdate),
+        }
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum ApUpdate {
+    Regular = 0,
+    Add = 1,
+    Add1 = 2,
+    // TODO: Check if this is correct
+    Add2,
+}
+
+impl TryFrom<&CairoMemoryCell> for ApUpdate {
+    type Error = InstructionDecodingError;
+
+    fn try_from(cell: &CairoMemoryCell) -> Result<Self, Self::Error> {
+        let flags = cell.value.limbs[3] >> FLAGS_OFFSET;
+        let ap_update = ((flags & AP_UPDATE_MASK) >> AP_UPDATE_OFF) as u8;
+
+        match ap_update {
+            0 => Ok(ApUpdate::Regular),
+            1 => Ok(ApUpdate::Add),
+            2 => Ok(ApUpdate::Add1),
+            4 => Ok(ApUpdate::Add2),
+            _ => Err(InstructionDecodingError::InvalidApUpdate),
+        }
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct CairoInstructionFlags {
+    pub opcode: CairoOpcode,
+    pub ap_update: ApUpdate,
+    pub pc_update: PcUpdate,
+    pub res_logic: ResLogic,
+    pub op1_src: Op1Src,
+    pub op0_reg: Op0Reg,
+    pub dst_reg: DstReg,
+}
```

Comment:
I'd put this at the beggining since it's the important thing, along with the rest of the instruction